### PR TITLE
gh #7131 Support for astype with literal strings

### LINKED ISSUE #7131: Array.astype not support with string literals
## Array.astype not working with literal strings

Providing a literal string like "float32" doesn't work with array.astype. Here is a reproducer from master:

```python
import numba
import numpy as np

@numba.njit
def ex():
      data = np.arange(1000)
      return data.astype("float32")
  
print(ex())
```
Which produces the output:

```python
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    print(ex())
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 421, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 360, in error_rewrite
    raise e
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 373, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 911, in compile
    cres = self._compiler.compile(args, return_type)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 84, in compile
    raise retval
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/dispatcher.py", line 107, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler.py", line 675, in compile_extra
    return pipeline.compile_extra(func)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler.py", line 419, in compile_extra
    return self._compile_bytecode()
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler.py", line 483, in _compile_bytecode
    return self._compile_core()
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler.py", line 462, in _compile_core
    raise e
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler.py", line 453, in _compile_core
    pm.run(self.state)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler_machinery.py", line 343, in run
    raise patched_exception
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler_machinery.py", line 334, in run
    self._runPass(idx, pass_inst, state)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler_machinery.py", line 289, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/compiler_machinery.py", line 262, in check
    mangled = func(compiler_state)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/typed_passes.py", line 105, in run_pass
    typemap, return_type, calltypes, errs = type_inference_stage(
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/typed_passes.py", line 83, in type_inference_stage
    errs = infer.propagate(raise_errors=raise_errors)
  File "/Users/nicholasriasanovsky/Documents/bodo/nick-numba/numba/core/typeinfer.py", line 1074, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of BoundFunction(array.astype for array(int64, 1d, C)) with parameters (Literal[str](float32))

During: resolving callee type: BoundFunction(array.astype for array(int64, 1d, C))
During: typing of call at test.py (7)


File "test.py", line 7:
def ex():
    <source elided>
    data = np.arange(1000)
    return data.astype("float32")
```

- [X] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

### PR DESCRIPTION
**Pull request by njriasan**
Adds support for Array.astype with string literal type names. This required two changes:
  1. Adding a lower_builtin for string literals.
  2. Fixes a bug in `get_call_type` where the literal option would never be tried if the unliteral version was preferred and returned None.

## Reference an existing issue

Closes #7131

### CODE REVIEW

**Comment by esc** [comment]
@njriasan thank you for submitting this to the Numba issue tracker. I have added it to the queue for review. Please note that we are currently in a burn-down period towards a release candidate, so the review for this PR may be somewhat delayed. We thank you in advance for your patience and understanding!

File: numba/core/types/functions.py [hunk 1]
```diff
@@ -405,7 +405,6 @@ def get_call_type(self, context, args, kws):
         literal_e = None
         nonliteral_e = None
         out = None
-
```
Comment (stuartarchibald) [inline numba/core/types/functions.py:408]: Please revert, this is a stylistic change, thanks.

File: numba/core/types/functions.py [hunk 2]
```diff
@@ -442,7 +441,10 @@ def get_call_type(self, context, args, kws):
                                 raise exc
                         nonliteral_e = exc
                     else:
-                        break
+                        # If out returns None don't break as we may need to use
+                        # a literal.
+                        if out is not None:
+                            break
```
Comment (stuartarchibald) [inline numba/core/types/functions.py:447]: I think the typing for the `@bound_function("array.astype")` should raise a `numba.core.errors.RequireLiteralValue` exception if a `unicode_type` appears, this would be consistent with how e.g. `@overload` would handle this situation and doesn't require another code path here.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, couple of things to resolve else looks good.

File: numba/core/types/functions.py (hunk 2 above)
Comment (njriasan) [reply numba/core/types/functions.py]: This change worked. I've reverted the changes to `numba/core/types/functions.py` and updated the bound function. Thank you.

File: numba/core/types/functions.py (hunk 1 above)
Comment (njriasan) [reply numba/core/types/functions.py]: Done.

**Comment by njriasan** [comment]
@stuartarchibald I've update this PR with your suggested changes. Thank you.

File: numba/core/typing/arraydecl.py [hunk 3]
```diff
@@ -445,9 +445,12 @@ def resolve_view(self, ary, args, kws):
 
     @bound_function("array.astype")
     def resolve_astype(self, ary, args, kws):
+        from numba.core.errors import RequireLiteralValue
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:448]: Please could you move this import to the top, thanks.

File: numba/core/typing/arraydecl.py [hunk 4]
```diff
@@ -445,9 +445,12 @@ def resolve_view(self, ary, args, kws):
...
         from .npydecl import parse_dtype
         assert not kws
         dtype, = args
+        if dtype == types.unicode_type:
+            raise RequireLiteralValue("array.astype if dtype is a string it must be constant")
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:453]: This should probably be explicitly tested with a non-constant string and the error message checked cf. https://github.com/numba/numba/blob/a8e24a98d0cce7464043f5eca061ea1f45c3f537/numba/tests/test_array_methods.py#L491-L494

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update, couple of minor things to resolve else looks good.

File: numba/core/types/functions.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: Thanks

File: numba/core/types/functions.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: Great, thanks.

File: numba/core/typing/arraydecl.py (hunk 3 above)
Comment (njriasan) [reply numba/core/typing/arraydecl.py:448]: Yes absolutely. Out of curiosity I thought I saw that other code segments do some imports within functions to reduce the overall import time. Is this correct and if so what is the methodology for importing inside a function.

Comment (njriasan) [reply numba/core/typing/arraydecl.py:448]: I understand that its clear in this example because the file already imports `numba.core.errors`.

Comment (stuartarchibald) [reply numba/core/typing/arraydecl.py:448]: Thanks. In general, unless proven to be an issue, imports go at the top, it makes refactoring and dealing with side effects easier.

**Comment by njriasan** [comment]
@stuartarchibald Sorry for the delay. I finally got the time to finalize this PR.

File: numba/core/typing/arraydecl.py (hunk 4 above)
Comment (njriasan) [reply numba/core/typing/arraydecl.py:452]: I added an additional test to check for this. Since the previous infrastructure lowers dtype as a constant, I had to provide a new check and run function. Hopefully this infrastructure change is acceptable even if there is only 1 test that runs.

**Comment by njriasan** [comment]
@stuartarchibald Just wanted to send you another ping in case the notification got lost in your inbox. No rush on this, but it should be a quick review when you have time.

File: numba/tests/test_array_methods.py [hunk 5]
```diff
@@ -465,12 +465,32 @@ def check(arr, dtype):
...
 
+        def run_dtype_arg(arr, dtype):
+            """
+            Equivalent to run but dtype is passed as
+            an arg instead of as a lowered constant.
+            This is used to check LiteralValue requirements.
+            """
+            pyfunc = lambda arr, dtype: arr.astype(dtype)
+            cres = self.ccache.compile(pyfunc, (typeof(arr), typeof(dtype)))
+            return cres.entry_point(arr, dtype)
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py:476]: This is not reached in the current tests.

File: numba/tests/test_array_methods.py [hunk 6]
```diff
@@ -465,12 +465,32 @@ def check(arr, dtype):
...
+        def check_dtype_arg(arr, dtype):
+            """
+            Equivalent to check but dtype is passed as
+            an arg instead of as a lowered constant.
+            This is used to check LiteralValue requirements.
+            """
+            expected = arr.astype(dtype).copy(order='A')
+            got = run_dtype_arg(arr, dtype)
+            self.assertPreciseEqual(got, expected)
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py:485]: This is not reached in the current tests.

File: numba/tests/test_array_methods.py [hunk 7]
```diff
@@ -491,6 +511,12 @@ def check(arr, dtype):
             check(arr, dt)
         self.assertIn('cannot convert from int32 to Record',
                       str(raises.exception))
+        # Non-Literal String
+        unicode_val = "float32"
+        with self.assertTypingError() as raises:
+            check_dtype_arg(arr, unicode_val)
+        self.assertIn('array.astype if dtype is a string it must be constant',
+                      str(raises.exception))
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py:519]: Suggest something like this as it's really only this one case that exists?:
```diff
diff --git a/numba/tests/test_array_methods.py b/numba/tests/test_array_methods.py
index 64c0eba..c130d88 100644
--- a/numba/tests/test_array_methods.py
+++ b/numba/tests/test_array_methods.py
@@ -469,25 +469,6 @@ class TestArrayMethods(MemoryLeakMixin, TestCase):
             got = run(arr, dtype)
             self.assertPreciseEqual(got, expected)
 
-        def run_dtype_arg(arr, dtype):
-            """
-            Equivalent to run but dtype is passed as
-            an arg instead of as a lowered constant.
-            This is used to check LiteralValue requirements.
-            """
-            pyfunc = lambda arr, dtype: arr.astype(dtype)
-            cres = self.ccache.compile(pyfunc, (typeof(arr), typeof(dtype)))
-            return cres.entry_point(arr, dtype)
-        def check_dtype_arg(arr, dtype):
-            """
-            Equivalent to check but dtype is passed as
-            an arg instead of as a lowered constant.
-            This is used to check LiteralValue requirements.
-            """
-            expected = arr.astype(dtype).copy(order='A')
-            got = run_dtype_arg(arr, dtype)
-            self.assertPreciseEqual(got, expected)
-
         # C-contiguous
         arr = np.arange(24, dtype=np.int8)
         check(arr, np.dtype('int16'))
@@ -515,10 +496,14 @@ class TestArrayMethods(MemoryLeakMixin, TestCase):
             check(arr, dt)
         self.assertIn('cannot convert from int32 to Record',
                       str(raises.exception))
-        # Non-Literal String
+
+        # Check non-Literal string raises
         unicode_val = "float32"
         with self.assertTypingError() as raises:
-            check_dtype_arg(arr, unicode_val)
+            @jit(nopython=True)
+            def foo(dtype):
+                np.array([1]).astype(dtype)
+            foo(unicode_val)
         self.assertIn('array.astype if dtype is a string it must be constant',
                       str(raises.exception))
```

File: numba/core/typing/arraydecl.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/core/typing/arraydecl.py:452]: Thanks, I've commented inline.

File: numba/core/typing/arraydecl.py [hunk 8]
```diff
@@ -448,6 +448,8 @@ def resolve_astype(self, ary, args, kws):
...
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:452]: This line needs wrapping to 80 chars.

File: numba/core/typing/arraydecl.py [hunk 9]
```diff
@@ -448,6 +448,8 @@ def resolve_astype(self, ary, args, kws):
         from .npydecl import parse_dtype
         assert not kws
         dtype, = args
+        if dtype == types.unicode_type:
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:451]:
```suggestion
        if isinstance(dtype, types.UnicodeType):
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update, I've left some comments inline, I think given the nature of the corner cases being tested being very explicit in the tests is preferable. Thanks again!

File: numba/core/typing/arraydecl.py [hunk 10]
```diff
@@ -448,6 +448,9 @@ def resolve_astype(self, ary, args, kws):
         from .npydecl import parse_dtype
         assert not kws
         dtype, = args
+        if isinstance(dtype, types.UnicodeType):
+            raise RequireLiteralValue("array.astype if dtype is a string \
+                 it must be constant")
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:453]:
```suggestion
            raise RequireLiteralValue(("array.astype if dtype is a string it "
                                       "must be constant"))
```

**Review by stuartarchibald — COMMENTED** [review]
This should fix CI.

**Comment by njriasan** [comment]
> This should fix CI.

Thank you!

**Comment by stuartarchibald** [comment]
@njriasan Congratulations on your first contribution to Numba!
